File: plugin/zeppelin/src/main/scala/jetbrains/zeppelin/idea/toolwindow/InterpretersView.scala
package jetbrains.zeppelin.idea.toolwindow

import java.awt.event.{ActionEvent, MouseAdapter, MouseEvent}

import com.intellij.openapi.Disposable
import com.intellij.openapi.project.Project
import com.intellij.ui.components.{JBList, JBScrollPane}
import javax.swing.{JComponent, JMenuItem, JPopupMenu, SwingUtilities}
import jetbrains.zeppelin.components.ZeppelinComponent

/**
  * Console that handle all zeppelin messages
  */
class InterpretersView(project: Project) extends JBScrollPane with Disposable {
  val innerList = new JBList[String]()

  object PopupItem extends Enumeration {
    type PopupItem = Value
    val RESTART_INTERPRETER: PopupItem.Value = Value("Restart")
    val SETTINGS: PopupItem.Value = Value("Settings")
    val SET_DEFAULT: PopupItem.Value = Value("Set default")
  }

  override def dispose(): Unit = {}

  this.setViewportView(innerList)
  innerList.getEmptyText.setText("Please, update the list of the interpreter")
  initPopupItemMenu()

  def getComponent: JComponent = {
    this
  }

  def getSelectedValue: String = innerList.getSelectedValue

  def updateInterpretersList(interpreters: Iterable[String]): Unit = {
    val interpretersArray = if (interpreters.nonEmpty) {
      val defaultElement = s"${interpreters.head} (default)"
      defaultElement +: interpreters.drop(1).toList.toArray
    } else {
      interpreters.toArray
    }
    val model = JBList.createDefaultListModel(interpretersArray: _*)
    innerList.setModel(model)
  }

  private def initPopupItemMenu(): Unit = {
    val refreshItem = new JMenuItem(PopupItem.RESTART_INTERPRETER.toString)
    refreshItem.addActionListener((e: ActionEvent) => popupElementAction(e))

    val settingsItem = new JMenuItem(PopupItem.SETTINGS.toString)
    settingsItem.addActionListener((e: ActionEvent) => popupElementAction(e))

    val makeDefaultItem = new JMenuItem(PopupItem.SET_DEFAULT.toString)
    makeDefaultItem.addActionListener((e: ActionEvent) => popupElementAction(e))

    val popupMenu = new JPopupMenu
    popupMenu.add(refreshItem)
    popupMenu.add(makeDefaultItem)
    popupMenu.add(new JPopupMenu.Separator())
    popupMenu.add(settingsItem)

    innerList.addMouseListener(new MouseAdapter() {
      override def mouseClicked(me: MouseEvent): Unit = {
        val index = innerList.locationToIndex(me.getPoint)
        innerList.setSelectedIndex(index)
        if (index == -1) return
        if (SwingUtilities.isRightMouseButton(me)) popupMenu.show(innerList, me.getX, me.getY)
        if (me.getClickCount == 2) setDefaultInterpreter()
      }
    })
  }

  private def openSettingsForm(): Unit = {
    ZeppelinComponent.connectionFor(project).service.openSettingsForm(getSelectedValue)
  }

  private def popupElementAction(e: ActionEvent): Unit = {
    val item: PopupItem.Value = PopupItem.withName(e.getActionCommand)
    item match {
      case PopupItem.RESTART_INTERPRETER => restartInterpreter()
      case PopupItem.SETTINGS => openSettingsForm()
      case PopupItem.SET_DEFAULT => setDefaultInterpreter()
      case _ => throw new Exception("Not implemented popup element")
    }
  }

  private def restartInterpreter(): Unit = {
    ZeppelinComponent.connectionFor(project).service.restartInterpreter(getSelectedValue)
  }

  private def setDefaultInterpreter(): Unit = {
    val interpreterName = getSelectedValue
    val connection = ZeppelinComponent.connectionFor(project)
    connection.service.setDefaultInterpreter(interpreterName)
    connection.updateInterpreterList()
  }
}

File: zeppelin/implementation/src/org/intellij/plugin/zeppelin/idea/toolwindow/InterpretersView.kt
package org.intellij.plugin.zeppelin.idea.toolwindow

import com.intellij.openapi.Disposable
import com.intellij.openapi.project.Project
import com.intellij.ui.components.JBList
import com.intellij.ui.components.JBScrollPane
import org.intellij.plugin.zeppelin.components.ZeppelinComponent
import org.intellij.plugin.zeppelin.models.ZeppelinException
import java.awt.event.ActionEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.JMenuItem
import javax.swing.JPopupMenu
import javax.swing.SwingUtilities

/**
 * Console that handle all zeppelin messages
 */
class InterpretersView(val project: Project) : JBScrollPane(), Disposable {
    val innerList = JBList<String>()

    enum class PopupItem(val value: String) {
        RESTART_INTERPRETER("Restart"),
        SETTINGS("Settings"),
        SET_DEFAULT("Set default"),
        SYNCHRONIZE("Synchronize")
    }

    override fun dispose() {}

    init {
        this.setViewportView(innerList)
        innerList.emptyText.text = "Please, update the list of the interpreter"
        initPopupItemMenu()
    }

    private fun getSelectedValue(): String = innerList.selectedValue

    fun updateInterpretersList(interpreters: List<String>) {
        val interpretersArray = if (interpreters.isNotEmpty()) {
            val defaultElement = "${interpreters.first()} (default)"
            listOf(defaultElement) + interpreters.drop(1)
        } else {
            interpreters
        }.toTypedArray()
        val model = JBList.createDefaultListModel(*interpretersArray)
        innerList.model = model
    }

    private fun initPopupItemMenu() {
        val popupMenu = JPopupMenu()
        PopupItem.values().forEach {
            val item = JMenuItem(it.value)
            item.addActionListener { action -> popupElementAction(action) }
            popupMenu.add(item)
        }

        innerList.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(me: MouseEvent) {
                val index = innerList.locationToIndex(me.point)
                innerList.selectedIndex = index
                if (index == -1) return
                if (SwingUtilities.isRightMouseButton(me)) popupMenu.show(innerList, me.x, me.y)
                if (me.clickCount == 2) setDefaultInterpreter()
            }
        })
    }

    private fun openSettingsForm() {
        ZeppelinComponent.connectionFor(project).service.openSettingsForm(getSelectedValue())
    }

    private fun popupElementAction(e: ActionEvent) {
        val item = PopupItem.values().find { it.value == e.actionCommand } ?: throw ZeppelinException(
                "Cannot parse an interpreter view menu element")

        when (item) {
            PopupItem.RESTART_INTERPRETER -> restartInterpreter()
            PopupItem.SETTINGS -> openSettingsForm()
            PopupItem.SET_DEFAULT -> setDefaultInterpreter()
            PopupItem.SYNCHRONIZE -> ZeppelinComponent.connectionFor(project).service.synchronizeInterpreter()
        }
    }

    private fun restartInterpreter() {
        ZeppelinComponent.connectionFor(project).service.restartInterpreter(getSelectedValue())
    }

    private fun setDefaultInterpreter() {
        val interpreterName = getSelectedValue()
        val connection = ZeppelinComponent.connectionFor(project)
        connection.service.setDefaultInterpreter(interpreterName)
        connection.updateInterpreterList(true)
    }
}